python/pyspark/sql/column.py

```diff
@@ -406,7 +406,13 @@ def substr(self, startPos, length):
         [Row(col=u'Ali'), Row(col=u'Bob')]
         """
         if type(startPos) != type(length):
-            raise TypeError("Can not mix the type")
+            raise TypeError(
+                "startPos and length must be the same type. "
+                "Got {startPos_t} and {length_t}, respectively."
```

Comment:
It needs to check the types in general and we need to hide the error message related with Java types. It is also true that we also need to make such logics in to Scala one to deduplicate this logic if they are duplicated. R has also a similar problem in some places. I don't think we should change this case anyway.

It looks we should ...

```
py4j.Py4JException: Method substr([class java.lang.Long ...
```

or we should introduce bridge methods in Scala side and implement this checking logic IIRC.